### ISSUE TITLE
Wav2vec2 insertion problem

### ISSUE THREAD

**Issue by Enescigdem** [speechbrain/speechbrain#911]
Hello, I used the wav2vec2 model from pretrained and I finetuned for my own data set which gives low WER scores in validation. But In test time, the model inserts a huge amount of meaningless insertions,espeacially for long utterances.
What could be wrong ? 
Thanks in advance

**Comment by TParcollet** [comment]
Interesting, you should not see differences in valid / test behavior if you have the same distribution of data. I don't know your recipe, but could you tell me if the validation and test searchers are the same? It most likely comes from them.

**Comment by Enescigdem** [comment]
Yes, i find this interesting.
The searcher i used in validation nnd test time is below.
beam_searcher: !new:speechbrain.decoders.S2SRNNBeamSearchLM
    embedding: !ref <emb>
    decoder: !ref <dec>
    linear: !ref <seq_lin>
    language_model: !ref <lm_model>
    lm_weight: !ref <lm_weight>
    ctc_linear: !ref <ctc_lin>
    bos_index: !ref <bos_index>
    eos_index: !ref <eos_index>
    blank_index: !ref <blank_index>
    min_decode_ratio: !ref <min_decode_ratio>
    max_decode_ratio: !ref <max_decode_ratio>
    beam_size: !ref <beam_size>
    eos_threshold: !ref <eos_threshold>
    using_max_attn_shift: !ref <using_max_attn_shift>
    max_attn_shift: !ref <max_attn_shift>
    temperature: !ref <temperature>

beam_size: 40 
lm_weight: 0.25
The other parameters are same in speechbrain/recipes .

**Comment by mravanelli** [comment]
Hi, are you using char or bpe as targets? In some cases (e.g., librispeech)
we see much better results when using chars rather than bpes with wav2vec.

**Comment by Enescigdem** [comment]
In my case , tokenizer type is unigram.
I will try char type  and share my new results as soon as possible.
Thank you so much

**Comment by mravanelli** [comment]
Great, please keep us updated!

**Comment by mravanelli** [comment]
This might be related to #924 as well. Are you using batch decoding? If so, it might be better sorting the sentences before creating the batches.

**Comment by Enescigdem** [comment]
Hi, 
When I used char type tokenizer, I faced out of vocab issues in validation time and I could not solve them by increasing the vocab size of the tokenizer.

**Comment by Enescigdem** [comment]
> This might be related to #924 as well. Are you using batch decoding? If so, it might be better sorting the sentences before creating the batches.

I always sort the datasets for every stage I use.

**Comment by Enescigdem** [comment]
I tested my char type tokenizer model with test set again Wer is about hundreds.

**Comment by TParcollet** [comment]
@Enescigdem if you are interested in switching to char you need to make two changes:
1. Train the tokenizer, and by looking at the logs or the generated.vocab files, you'll get the size of your vocabulary. 
2. set your output_neuron number to this value + x (with x depending on if you are using eos sos tokens etc).

For the discrepancy between validation and test, I propose you to do something quite unusual (and incorrect) -> Train your tokenizer and the dev and test data. Then redo the training and evaluate. It's just to see if there is, for some reasons, a huge discrepancy in the vocabulary. 

How much hours are we talking about here ? (used to train the tokenizer). This kind of errors may appear with very very very small training datasets for the tokenizer.

**Comment by Adel-Moumen** [comment]
Hello @Enescigdem,

Is there any news on this issue, please? Many thanks!

**Comment by Adel-Moumen** [comment]
Hello,

It seems there has been no activity for a long time, and that @TParcollet has answered your problem, so I am closing this issue.

Feel free to reopen it if you still encounter the problem. Thank you. 🙂
